### PR TITLE
Uploading files

### PR DESCRIPTION
**Pull request by haxxorsid**
### Summary of Changes

* Added Upload file dialog.
* User can upload multiple files, all selected files are listed in ListBox
* User can see errors related file upload(eg: file already exists, unsupported ext) in the upload-file-dialog.

### Improvements needed
* UploadFileDialog.tsx can be improved.
* Possible improvements are stated below in reviews.

### CODE REVIEW

File: src/components/UploadFileDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+      const name: string = file.name;
+      const fileType: string = filetypeForExtension(name.split(".").pop());
+      let data: any;
+      try {
+        data = await this._readUploadedFileAsText(file);
+        this.setState(prevState => ({
+          files: [...prevState.files, {name, fileType, data}]
+        }));
+      } catch (e) {
+          console.log("Unable to read the file!");
```
Comment (haxxorsid) [inline src/components/UploadFileDialog.tsx]: This needs to be handled like stuff in _nameError(). Idk how should I connect this with output of _nameError.

File: src/components/UploadFileDialog.tsx [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+  async _handleUpload(files: FileList) {
+    this.setState({files: []});
+    Array.from(files).forEach(async (file: any) => {
+      const name: string = file.name;
+      const fileType: string = filetypeForExtension(name.split(".").pop());
+      let data: any;
+      try {
+        data = await this._readUploadedFileAsText(file);
+        this.setState(prevState => ({
+          files: [...prevState.files, {name, fileType, data}]
```
Comment (haxxorsid) [inline src/components/UploadFileDialog.tsx]: `data `of file is added in state even if _nameError function gives some errors. So data must be added later. I need solution for adding data in state later.

File: src/components/App.tsx [hunk 3]
```diff
@@ -730,6 +733,27 @@ export class App extends React.Component<AppProps, AppState> {
...
+        <UploadFileDialog
+          isOpen={true}
+          directory={this.state.uploadFileDialogDirectory}
+          onCancel={() => {
+            this.setState({ uploadFileDialogDirectory: null });
+           }}
+          onUpload={(files: File[]) => {
+            if (!(this.state.uploadFileDialogDirectory instanceof Project)) {
+              files.map((file: File) => {
+                this.state.uploadFileDialogDirectory.children.push(file);
```
Comment (haxxorsid) [inline src/components/App.tsx]: Oh. Maybe I should use addFile function here instead of pushing?

File: src/components/UploadFileDialog.tsx [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    this.state = {
+      files: []
+    };
+  }
+  _nameError() {
+    const directory = this.props.directory;
+    let errorStr: string = "";
+    this.state.files.map((file: any) => {
+      if (file.fileType == null) {
+          errorStr += `File '${file.name}' is not supported.`;
```
Comment (mbebenita) [inline src/components/UploadFileDialog.tsx]: Indentation.

File: src/components/UploadFileDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+              this.props.onCancel();
+            }}
+          />
+          <Button
+            icon={<GoFile />}
+            label="Upload"
+            title="Upload New File"
+            isDisabled={!this.state.files.length || !!this._nameError()}
+            onClick={() => {
+            const newFiles: File[] = [];
```
Comment (mbebenita) [inline src/components/UploadFileDialog.tsx]: Indentation.

File: src/components/UploadFileDialog.tsx (hunk 1 above)
Comment (mbebenita) [reply src/components/UploadFileDialog.tsx]: Can you associate errors with each file structure in `state.files`, and maybe show errors inside of the list items?

**Comment by mbebenita** [comment]
@yurydelendik is it possible to upload directories?

**Comment by mbebenita** [comment]
@haxxorsid instead showing files using a list box, we could embed a DirectoryTree view in the upload file dialog. Instead of keeping track of files as an array of files, you could create a Directory and stick the files in there, and that's the object you'd pass to the embedded DirectoryTree. To prevent users from right clicking on files in the embedded DirectoryTree component, we can add a prop to it that disables that functionality.

**Comment by mbebenita** [comment]
Lastly, we can probably add a button at the bottom after Cancel that is styled like the other buttons, which clicks on the upload input element.

image

**Comment by yurydelendik** [comment]
> @yurydelendik is it possible to upload directories?

https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement lists `allowdirs` and `webkitdirectory` properties -- I guess it's possible

File: src/components/UploadFileDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+  }
+  _nameError() {
+    const directory = this.props.directory;
+    let errorStr: string = "";
+    this.state.files.map((file: any) => {
+      if (file.fileType == null) {
+          errorStr += `File '${file.name}' is not supported.`;
+      }
+      if (file.name) {
+        if (!/^[a-z0-9\.\-\_]+$/i.test(file.name)) {
```
Comment (yurydelendik) [inline src/components/UploadFileDialog.tsx]: Extract regexp check as a function with meaningful name, e.g. `isFilenameValid`

File: src/components/UploadFileDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+      } catch (e) {
+          console.log("Unable to read the file!");
+      }
+    });
+  }
+  _readUploadedFileAsText(inputFile: any) {
+    const temporaryFileReader = new FileReader();
+    return new Promise((resolve, reject) => {
+      temporaryFileReader.onerror = () => {
+      temporaryFileReader.abort();
```
Comment (yurydelendik) [inline src/components/UploadFileDialog.tsx]: nit: bad indent

**Comment by mbebenita** [comment]
> I guess it's possible

Cool, so this means we could recursively upload them and display them into a DirectoryTree component, maybe with some indicators like (this file already exists, etc).

**Comment by mbebenita** [comment]
@haxxorsid let's handle the non directory case for now, and switch to the DirectoryTree view, we can do the rest later.

File: src/components/UploadFileDialog.tsx (hunk 7 above)
Comment (mbebenita) [reply src/components/UploadFileDialog.tsx]: Not sure why tslint is not reporting these errors.

**Comment by haxxorsid** [comment]
"Lastly, we can probably add a button at the bottom after Cancel that is styled like the other buttons, which clicks on the upload input element." 
You mean it will open browser's upload file dialog?

**Comment by mbebenita** [comment]
> it will open browser's upload file dialog

Yes, that's possible. @yurydelendik has some tricks.

**Comment by haxxorsid** [comment]
@mbebenita 
Assume: I have this <input id='uploadfile' type='file' multiple hidden/>
I do this in Button Component:
```
onClick={() => {
  document.getElementById('uploadfile').click();
}}
```
Cool?

**Comment by mbebenita** [comment]
For some reason I can't see this comment in github, so replying over
e-mail. But yes, you should be able to do that, and then hide the
uploadfile element with CSS.

File: src/components/UploadFileDialog.tsx [hunk 8]
```diff
@@ -128,16 +137,24 @@ export class UploadFileDialog extends React.Component<UploadFileDialogProps, {
           <Button
             icon={<GoFile />}
             label="Upload"
-            title="Upload New File"
-            isDisabled={!this.state.files.length || !!this._nameError()}
+            title="Upload files"
+            onClick={() => {
+              document.getElementById("file-upload-input").click();
```
Comment (mbebenita) [inline src/components/UploadFileDialog.tsx]: This is a React anti-pattern. I think what you need to do, is add a `ref` on the FileUploadInput component, then here you can say something like `this.fileUploadInput.open()`.

File: src/components/Widgets.tsx [hunk 9]
```diff
@@ -85,22 +84,7 @@ export class FileUploadInput extends React.Component<{
...
-            <div className="column">
-              <div>{input}</div>
-            </div>
-          </div>
-          {this.props.error && <div className="text-input-box-error">{this.props.error}</div>}
-        </div>;
-    } else {
-      return input;
-    }
+    return <input id="file-upload-input" type="file" onChange={this.props.onChange} multiple hidden/>;
```
Comment (mbebenita) [inline src/components/Widgets.tsx]: Here you can do a `ref` to get a handle to the input element. Then add a function called `open()` that files the click event on it, which is called from a parent component.

**Comment by haxxorsid** [comment]
@mbebenita Yep.

**Comment by haxxorsid** [comment]
@mbebenita I've fixed everything including indents. Euhm I don't know what is blocking this more.

**Comment by mbebenita** [comment]
@haxxorsid thanks!

**Pull request merged**